Source/WTF/wtf/SystemTracing.h

```diff
@@ -195,72 +200,109 @@ using WTF::tracePoint;
+// but is not enabled by tailspin. This means we generally emit signposts only if a developer is
+// actively profiling their system.
+#define WTFEmitSignpost(pointer, name, ...) WTFEmitSignpostWithTimeDelta((pointer), name, Seconds(), ##__VA_ARGS__)
+#define WTFBeginSignpost(pointer, name, ...) WTFBeginSignpostWithTimeDelta((pointer), name, Seconds(), ##__VA_ARGS__)
+#define WTFEndSignpost(pointer, name, ...) WTFEndSignpostWithTimeDelta((pointer), name, Seconds(), ##__VA_ARGS__)
+
+// These work like WTF{Emit,Begin,End}Signpost, but offset the timestamp associated with the event
+// by timeDelta (which should of type WTF::Seconds).
+#define WTFEmitSignpostWithTimeDelta(pointer, name, timeDelta, ...) \
+    WTFEmitSignpostWithType(WTFOSSignpostTypeEmitEvent, os_signpost_event_emit, (pointer), name, (timeDelta), " %{signpost.description:event_time}llu", "" __VA_ARGS__)
+
+#define WTFBeginSignpostWithTimeDelta(pointer, name, timeDelta, ...) \
+    WTFEmitSignpostWithType(WTFOSSignpostTypeBeginInterval, os_signpost_interval_begin, (pointer), name, (timeDelta), " %{signpost.description:begin_time}llu", "" __VA_ARGS__)
+
+#define WTFEndSignpostWithTimeDelta(pointer, name, timeDelta, ...) \
+    WTFEmitSignpostWithType(WTFOSSignpostTypeEndInterval, os_signpost_interval_end, (pointer), name, (timeDelta), " %{signpost.description:end_time}llu", "" __VA_ARGS__)
+
+#define WTFEmitSignpostWithType(type, emitMacro, pointer, name, timeDelta, timeFormat, format, ...) \
+    do { \
+        if (UNLIKELY(kdebug_is_enabled(KDBG_EVENTID(DBG_APPS, DBG_APPS_WEBKIT_MISC, 0)))) { \
+            if (WTFSignpostIndirectLoggingEnabled) \
+                WTFEmitSignpostIndirectlyWithType(type, pointer, name, timeDelta, format, ##__VA_ARGS__); \
+            else { \
+                Seconds delta = (timeDelta); \
+                if (delta) \
+                    WTFEmitSignpostDirectlyWithType(emitMacro, pointer, name, format timeFormat, ##__VA_ARGS__, WTFCurrentContinuousTime(delta)); \
+                else \
+                    WTFEmitSignpostDirectlyWithType(emitMacro, pointer, name, format, ##__VA_ARGS__); \
+            } \
+        } \
+    } while (0)
+
+#define WTFEmitSignpostDirectlyWithType(emitMacro, pointer, name, format, ...) \
+    do { \
+        os_log_t wtfHandle = WTFSignpostLogHandle(); \
+        const void *wtfPointer = (pointer); \
+        os_signpost_id_t wtfSignpostID = wtfPointer ? os_signpost_id_make_with_pointer(wtfHandle, wtfPointer) : OS_SIGNPOST_ID_EXCLUSIVE; \
+        emitMacro(wtfHandle, wtfSignpostID, #name, format, ##__VA_ARGS__); \
+    } while (0)
+
+#define WTFEmitSignpostIndirectlyWithType(type, pointer, name, timeDelta, format, ...) \
+    os_log(WTFSignpostLogHandle(), "type=%d name=%d p=%" PRIuPTR " ts=%llu " format, type, WTFOSSignpostName ## name, reinterpret_cast<uintptr_t>(pointer), WTFCurrentContinuousTime(timeDelta), ##__VA_ARGS__)
```

Comment:
Another option here could perhaps be to delegate this to a signpost observer, that would send a signpost message to the Networking process. Then you could avoid the sscanf in the Networking process.

Reply:
I don't actually see a way to do this without invoking `os_log`. That's because `os_signpost` takes an `os_log`-compatible format string and I don't see any API that can actually expand that format string outside of calling `os_log` and hooking into the object passed to `os_log_hook`.